### DISCUSSION TITLE
How to compute optimal receptive field for SLEAP

### DISCUSSION THREAD

**Discussion by catubc** [talmolab/sleap#2210] [General]
Hello
We've got top down mouse camera vids with pretty high resolution, 6-8 pixels/mm. The videos are 1024 x 746 pixels.

 We're finding that SLEAP completely fails to learn with default settings.  

We made the receptive field bigger and it's starting to learn now.  

Is there some calculation for this that we missed?  As you can see in the vids, our mice can be as much as 400 pixels across just for the body, excluding the tail.  

ChatGPT seems to suggest:
- Centroid model: RF ≈ 450–500 px
- Confidence map model (for body parts): RF ≈ 150–250 px

Full mouse (centroid model)
-  5 down blocks with 2–3 convs/block
- down_blocks = 5, convs_per_block = 3 → RF ≈ 475–500 px
- OR max_stride = 5

Smaller receptive field (e.g. paw or ear)
-  3–4 down blocks, 2–3 convs/block
- down_blocks = 4, convs_per_block = 2 → RF ≈ 220 px
- OR max_stride = 3 or 4

Thanks so much in advance.  

The video shows SLEAP on the left (with already larger RFs than default) - and YOLO on right (with identical training sets).

![output2](https://github.com/user-attachments/assets/f1256772-f371-4c7f-bba6-aba58a0c816b)

**Comment by eberrigan** [comment]
Hello @catubc ,

We have a feature in the GUI training window to assist with parameter choices. 

Likely you can reduce the input scaling to 0.5 or even 0.25 since your images are such high resolution. Then you can adjust the max stride so the receptive field fits your mouse for the centroid model. 

For the centered instance model you will leave the input scaling at 1.0 and should adjust the max stride so that you can easily see keypoints and their contexts on the instance-centered crops fed into this model.

Please see the documentation here for more explanation https://sleap.ai/develop/guides/choosing-models.html#choosing-models.

So the SLEAP model is detecting and the YOLO model is not currently?

Best,

Elizabeth

**Comment by sronilsson** [comment]
Hi @eberrigan thanks for this, I appreciate it! am trying these settings right now. I have three questions below it would be great to get your help with. 

I have to go down to `input scaling` 0.1 in the centroid model and  `max stride` 128 in the center instance model to get rectangular bounding boxes in the preview window that encompass the animal, including its tail. As these values are lower than the suggested 0.5-0.20, do you see any issues with these values?

![image](https://github.com/user-attachments/assets/743186aa-9ad4-49db-b335-19eb27499305)

When starting the training I get pop-ups labeled with green/red key points, like below: are these labels supposed to be the labeled centroids? I can see that these locations would be the centroid of the axis aligned bounding box around the animal, including its tail, but it’s not really the centroid of the animal hull. Is there anything in the GUI I could the make this location the center of the animal hull and would this improve the model? 

![image](https://github.com/user-attachments/assets/7bf19e97-60d1-498e-ab33-ee011c49c3a9)
![image](https://github.com/user-attachments/assets/44faf4cf-3ea5-44c0-a6d5-d0dc68cbb95d)

**yes SLEAP is definitely outperforming YOLO.**  The reason we tested YOLO is the huge amounts of data that must be analysed and latency is a priority. The readme suggests 600+ FPS, is there anything in terms of GUI settings I should think about to get closer to this value?

**Reply by talmo** [reply]
> When starting the training I get pop-ups labeled with green/red key points, like below: are these labels supposed to be the labeled centroids? I can see that these locations would be the centroid of the axis aligned bounding box around the animal, including its tail, but it’s not really the centroid of the animal hull. 

Yep we just compute the midpoint of the bounding box, so it's often better to just select a specific landmark as the anchor point so it's more reliably yoked to a part of the anatomy of the animal.

> The readme suggests 600+ FPS, is there anything in terms of GUI settings I should think about to get closer to this value?

Basically to hit 600+ FPS you'll want to export to TensorRT. There's more detail in the paper if you search for the inference speed methods.

Some resources on speed:

https://github.com/talmolab/sleap/discussions/1276#discussioncomment-5649963

https://github.com/talmolab/sleap/discussions/959#discussioncomment-3655760

**Comment by eberrigan** [comment]
For centroid:
That should be fine since your mouse is visible at that scale. You could try different combinations of input scaling and max stride that fit your mouse and see if you get better model evaluation metrics. 

For centered-instance:
You do not need the receptive field size to encompass the entire mouse for the centered instance model. The receptive field for the centered instance model needs to encompass the keypoints and their context. You could try max stride of 64 and see if you get the same or better results.

Good luck!
